### PR TITLE
Classify unusable schema-designer responses in create-sql-schema (#1322)

### LINKED ISSUE #1322: `create-sql-schema` fails with "Error reading JObject from JsonReader" on valid input
**Severity:** blocks the SQL-script path entirely.

### Steps

```json
{ "command": "create-sql-schema",
  "args": { "environment-name": "<env>",
            "schema-name": "LabTempSqlFix",
            "package-name": "Custom",
            "caption": "Temporary data fix" } }
```

### Expected

A SQL script schema is created in the target package (per the tool contract, which lists only
`schema-name` and `package-name` as required).

### Actual

```text
Error reading JObject from JsonReader. Path '', line 0, position 0.
```

Reproduced twice, including with `caption` and `description` omitted so that only the two required
parameters were sent. The error appears to come from parsing an empty or non-JSON response body
from `ScriptSchemaDesignerService`, and gives the caller nothing to act on — no HTTP status, no
service name, no response excerpt.

### Suggested correction

Detect a non-JSON / empty service response and surface the status code and a response excerpt
instead of the raw deserializer message.

---

---

Environment common to all reports:

- clio knowledge library `com.creatio.clio` version **1.13.36**, sequence `1013036000`, bundle digest `8654a2af4df0c86a8a58d6483d1b3f023a6384118aca21ca7aa00dc2cdc0f9b5`
- Creatio **10.0.0.858**, .NET Framework 4.8, PostgreSQL, cloud-hosted
- `cliogate` 2.0.0.46
- Called through the clio MCP server via the `clio-run` generic executor

### PR DESCRIPTION
**Pull request by Alexandr-Kravchuk**
🔗 **Issue:** [#1322](https://github.com/Advance-Technologies-Foundation/clio/issues/1322)

## Root cause

`ServiceModel/ScriptSchemaDesignerService.svc` is not routed on the affected platform builds. Every method of that route answers `HTTP 404` with `Content-Length: 0`.

`IApplicationClient.ExecutePostRequest` does not expose the HTTP status — a 404 with an empty body comes back as an empty string, indistinguishable from a served endpoint that answered 200 with no body. `JObject.Parse("")` then throws, and the command's catch-all reported the raw deserializer text as the whole error:

```
Error reading JObject from JsonReader. Path '', line 0, position 0.
```

That message names neither the service nor the route, so a missing platform endpoint read as a clio JSON bug and the reporter retried a create that could never succeed.

## Fix

- `ServiceResponseJsonGuard.TryParseJObject` — the Newtonsoft counterpart of the existing `Deserialize` guard — now backs **all 6** `JObject.Parse` sites in `SchemaDesignerHelper`. A failure names the service and operation (for example `ScriptSchemaDesignerService CreateNewSchema`), the URL, and a locally authored hint. An HTML error page or login page is classified without echoing its markup.
- The empty-body message is honest about the unknown HTTP status instead of claiming the server accepted the request (see #1317).
- `SaveSchema` now reports an `outcomeUnknown` outcome: `create-sql-schema` re-reads the schema and reports it as found, absent, or explicitly unverified, rather than guessing.
- The duplicate-name check resolves through `ResolveSchemaUId` instead of `SchemaNameExists`, so a `SelectQuery` that could not be answered aborts instead of reading as "absent".

## Tests

```
dotnet test clio.tests/clio.tests.csproj --filter "Category=Unit&(Module=Command|Module=Package)"
dotnet test clio.mcp.e2e/clio.mcp.e2e.csproj -f net10.0 --filter "FullyQualifiedName~SqlSchemaCreateNonJsonResponseE2ETests"
```

The 6 failures in the `Module=Command|Module=Package` run are the known macOS-only baseline, present on a clean `origin/master` worktree as well.

## Verification on a live stand

Creatio 10.1.725, .NET Framework. Before the change, `create-sql-schema` returned only the JSON parser message above. After it, the failure names `ScriptSchemaDesignerService CreateNewSchema`, the URL, and the hint that the route is not served by this platform build.

The success path is **not** verifiable on this stand: the SQL-script designer service does not exist there — no SQL-script manager appears among the stand's `ManagerName` values, so the schema type is absent from that platform build entirely. Proxy evidence for the shared code path comes from `create-schema` on the `SourceCode` path, where `SourceCodeSchemaDesignerService.svc/CreateNewSchema` answers HTTP 200 with a schema payload through the same guarded parse.

## Reviews

- Docs reviewed, updated (`clio/help/en/create-sql-schema.txt` + `clio/docs/commands/create-sql-schema.md`); no option changed.
- MCP reviewed. The resident tool `[Description]` is unchanged deliberately — it counts against the `tools/list` budget — and e2e coverage was added.
- ClioRing compatibility reviewed, no Ring-consumed contract changed (verified by grep for `create-sql-schema` / `SqlSchemaCreate` across `clio-ring/`: no hits).

## Follow-ups

- `SourceCodeSchemaCreate` has the same `SchemaNameExists` gap and should get the same treatment.
- Determining the correct SQL-script designer route on platforms that do have it belongs to #1317.

Closes #1322

### CODE REVIEW

File: clio/Command/SchemaDesignerHelper.cs [hunk 1]
```diff
@@ -125,7 +151,9 @@ private static (string uId, string error) ResolveSchemaUIdSingle(
 		var query = BuildSelectUIdByName(schemaName, kind.ManagerName);
 		string url = urlBuilder.Build(SelectQueryRoute);
 		string responseJson = client.ExecutePostRequest(url, query.ToString(Formatting.None));
-		JObject selectResponse = JObject.Parse(responseJson);
+		(JObject selectResponse, string parseError) = ParseServiceResponse("SelectQuery", url, responseJson);
+		if (parseError != null)
+			return (null, parseError);
```
Comment (Alexandr-Kravchuk) [inline clio/Command/SchemaDesignerHelper.cs]: **Major (bug-risk):** SchemaNameExists now swallows an unparseable SelectQuery body, so SourceCodeSchemaCreate proceeds to create where it previously aborted.

Before this PR ResolveSchemaUIdSingle called JObject.Parse directly; an empty/HTML SelectQuery body threw, escaped SchemaNameExists, and SourceCodeSchemaCreate.TryCreate's catch-all stopped the command before any write. Now the same body becomes `(null, parseError)` here; SchemaNameExists (unchanged: `(string uId, _) = ResolveSchemaUId(...); return uId != null;`) discards the error and returns false, and `clio/Command/SourceCodeSchemaCreate.cs:82` proceeds to CreateNewSchema -> SaveSchema. A transport failure (expired session answering a login page, transient empty body) is read as "schema does not exist" — the hazard the comment in SqlSchemaCreate.cs names and fixes for the SQL-script command, introduced here for the source-code command. Whether SourceCodeSchemaDesignerService.SaveSchema rejects a same-name save was not verified against the platform, hence Major rather than Blocker.

Binding: D-2 — the duplicate-name check must abort when SelectQuery is unanswerable instead of reading as "absent".

Fix: SchemaNameExists has one remaining caller (SourceCodeSchemaCreate.cs:82). Either make it stop swallowing errors (hard failure for any error that is not IsSchemaNotFound) or delete it and make SourceCodeSchemaCreate use the same ResolveSchemaUId + IsSchemaNotFound branch SqlSchemaCreate now uses. Cover it with a SourceCodeSchemaCreate unit test asserting CreateNewSchema is not called when the duplicate-check SelectQuery body is empty.

File: clio/Command/SchemaDesignerHelper.cs [hunk 2]
```diff
@@ -274,7 +327,10 @@ internal static (JObject schema, string error) LoadSchema(
 		};
 		string designerUrl = urlBuilder.Build(kind.GetRoute);
 		string json = client.ExecutePostRequest(designerUrl, request.ToString(Formatting.None));
-		JObject response = JObject.Parse(json);
+		(JObject response, string parseError) = ParseServiceResponse(
```
Comment (Alexandr-Kravchuk) [inline clio/Command/SchemaDesignerHelper.cs:386]: **Major (testing):** the rerouted LoadSchema/SaveSchema/EnumerateSchemaLayers parse sites have no test for the changed behavior.

The diff inserts ParseServiceResponse at LoadSchema, EnumerateSchemaLayers, EnumerateSchemaLayersBatch and both SaveSchema overloads, but the new tests touch only CreateNewSchemaUrl/SaveSchemaUrl via SqlSchemaCreateCommand and the single SelectQuery resolve. LoadSchema (GetSchema), the enumerators and the 4-arg SaveSchema used by SqlSchemaUpdate.cs:96, SourceCodeSchemaUpdate.cs:74 and ClientUnitSchemaOptions.cs:168 have no empty/HTML-body case: SqlSchemaUpdateCommandTests and SqlSchemaGetCommandTests contain no string.Empty or `<html` stub on the GetSchema/SaveSchema URLs. No GitHub lane runs clio.mcp.e2e and the TeamCity MCP e2e is advisory, so unit mirrors are the only merge gate for this behavior. (No tests were executed; caller observations come from the master-based worktree.)

Binding: AC-1 — classified failure instead of raw Newtonsoft text at every guarded parse site.

Fix: in SqlSchemaGetCommandTests return string.Empty and an HTML page from the GetSchema URL and assert the error contains "ScriptSchemaDesignerService GetSchema" and the URL and not "Error reading JObject"; in SqlSchemaUpdateCommandTests return string.Empty from the SaveSchema URL and assert the classified SaveSchema error surfaces through the 4-arg overload; add one EnumerateSchemaLayers case with an empty SelectQuery body asserting the "SelectQuery" label.

**Review by Alexandr-Kravchuk — COMMENTED** [review]
**Decision: request changes** (posted as a comment — GitHub does not allow REQUEST_CHANGES on the author's own pull request).

Request changes. The guard behind the six JObject.Parse sites delivers AC-1/AC-2 cleanly, but the same change makes SchemaNameExists swallow an unparseable SelectQuery body, so create-source-code-schema now proceeds to create where it used to abort — the hazard the PR fixes for SQL schemas is introduced for its sibling (Major, high, objective). Four of six rerouted parse sites have no test of the changed behavior, and IsSchemaNotFound branches on unpinned local prose. Minors: lossy 4-arg SaveSchema overload, test naming, brittle hint assertion. No build/tests/CI were run; caller observations come from a master-based worktree.

**Requirement Coverage**

- AC-1 (classified failure instead of raw Newtonsoft text): covered — guard behind all six JObject.Parse sites; tests cover only the create-sql-schema path, remaining sites untested.
- AC-2 (name service/operation/URL): covered.
- AC-3 (HTTP status code): unclear — not reachable through IApplicationClient.ExecutePostRequest, message states status unknown, deferred to #1317.
- AC-4 (response excerpt): covered for non-HTML bodies; deliberately omitted for HTML/empty bodies.
- AC-5 (schema actually created): out of scope — route absent on measured Creatio build, deferred to #1317.
- D-1 (read-back after unusable SaveSchema answer): covered for create-sql-schema; four sibling callers keep the lossy 4-arg overload.
- D-2 (duplicate check aborts when unanswerable): covered for create-sql-schema, but regressed for create-source-code-schema via SchemaNameExists.

**Other findings (not posted inline)**

- Major (medium confidence) — `clio/Command/SchemaDesignerHelper.cs:303`: IsSchemaNotFound keys control flow off locally authored error prose produced in three places (SchemaDesignerHelper.cs:108, :137, GetClassicPageSourcesCommand.cs:623) with no shared constant; a wording change silently flips create-sql-schema to "Could not check whether schema already exists". The sibling `metadata is missing UId` text passes the prefix but not the substring, so an existing schema with a blank UId is reported as an unanswerable check. Neither classification is pinned by a test. Fix: return a discriminated outcome from ResolveSchemaUId (or one shared const for the not-found text used by all producers and the matcher), decide explicitly how `metadata is missing UId` is classified, and add a TryCreate case where the second SelectQuery returns `{"success":false,"errorInfo":{"message":"Schema not found in cache"}}` asserting abort and no CreateNewSchema post.
- Minor — `clio/Command/SchemaDesignerHelper.cs:353`: the 4-argument SaveSchema overload discards the new outcomeUnknown signal for SqlSchemaUpdate.cs:96, SourceCodeSchemaUpdate.cs:74, SourceCodeSchemaCreate.cs:101 and ClientUnitSchemaOptions.cs:168, so those commands still report an empty/HTML save answer as an observed failure. Fix: remove the overload and pass `out bool outcomeUnknown` at each caller, or mark it obsolete and list the four callers in the tracked follow-up issue.
- Minor — `clio.tests/Command/SqlSchemaCreateCommandTests.cs:151`: the ten new tests (TryCreate_Reports_..., TryCreate_Verifies_..., TryCreate_Aborts_..., TryParseJObject_ReportsEmptyBody_WithOperationAndUrl onward) do not follow the mandatory MethodName_ShouldExpectedBehavior_WhenCondition naming from project-context.md; e.g. TryCreate_ShouldAbort_WhenDuplicateCheckCannotBeAnswered.
- Minor — `clio.tests/Command/SqlSchemaCreateCommandTests.cs:170`: `.And.Contain("unlocked", …)` couples the assertion to one word of DesignerServiceHint; assert `SchemaDesignerHelper.DesignerServiceHint` instead.

Suppressed by review policy: 1 non-high-confidence Major(s) kept out of inline comments.

File: clio/Command/SchemaDesignerHelper.cs (hunk 1 above)
Comment (Alexandr-Kravchuk) [reply clio/Command/SchemaDesignerHelper.cs]: Good catch — fixed in b40155be7. `SchemaNameExists` is gone; `create-source-code-schema` now calls `ResolveSchemaUId` directly and branches on the new discriminated `SchemaResolveResult` (Resolved / NotFound / Unanswerable), exactly like `create-sql-schema`. Only an answered `NotFound` licenses the create — an empty/HTML body or a DataService failure envelope aborts. A blank UId is classified Unanswerable too, since the row existing is not evidence the schema is absent. Covered by `TryCreate_ShouldAbort_WhenDuplicateCheckCannotBeAnswered` in `SourceCodeSchemaCreateCommandTests`, which asserts `CreateNewSchema` is never posted.

File: clio/Command/SchemaDesignerHelper.cs (hunk 2 above)
Comment (Alexandr-Kravchuk) [reply clio/Command/SchemaDesignerHelper.cs:386]: Agreed — added in b40155be7. `SqlSchemaGetCommandTests` now covers the LoadSchema site with empty-body and HTML-page cases (the HTML one also asserts the markup is never echoed back), `SqlSchemaUpdateCommandTests` covers the SaveSchema site with an empty body and asserts the unknown-outcome note, and `SchemaDesignerHelperTests` covers `EnumerateSchemaLayers` with an empty SelectQuery body. All three assert the raw `Error reading JObject` text does not leak.

**Comment by Alexandr-Kravchuk** [comment]
Pushed b40155be7 covering the summary-only findings as well:

- **Discriminated resolve outcome.** `IsSchemaNotFound` (string matching on the error prose) is gone. `ResolveSchemaUId` now returns `SchemaResolveResult` with `Resolved` / `NotFound` / `Unanswerable`, so server-authored prose can never pick a control-flow branch. "Metadata is missing UId" is explicitly classified `Unanswerable`, not `NotFound`. New test `TryCreate_ShouldAbort_WhenDuplicateCheckReturnsFailureEnvelope` feeds `{"success":false,"errorInfo":{"message":"Schema not found in cache"}}` and asserts the create aborts.
- **Lossy overload removed.** The 4-arg `SaveSchema` is deleted; every caller (sql/source-code/client-unit update, source-code create) now takes `out bool outcomeUnknown` and appends `SchemaDesignerHelper.SaveOutcomeUnknownNote` so an unobserved write is never reported as an observed rejection.
- **Test naming.** New and touched tests renamed to `MethodName_ShouldExpectedBehavior_WhenCondition`.
- **Hint assertion.** The empty-`CreateNewSchema` test now asserts the whole `SchemaDesignerHelper.DesignerServiceHint` instead of the single word "unlocked".
- **URL segment / redaction.** The URL is emitted as its own trailing segment (`... . URL: <url> Next sentence`) instead of inside parentheses, so MCP redaction to `[redacted-uri]` no longer swallows the closing bracket or period. Rationale recorded in `docs/knowledge/McpServer/uri-redaction-swallows-punctuation-after-a-url.md`.
- **Non-object JSON bodies.** `ServiceResponseJsonGuardTests` now covers `null`, `[]` and `123`, asserting the message opens with the operation and names the URL rather than leaking raw Newtonsoft text.

Validated locally: `dotnet test clio.tests/clio.tests.csproj --filter "Category=Unit&(Module=Command|Module=Package)"` (only the 6 known macOS-baseline failures) and `dotnet test clio.mcp.e2e --filter "FullyQualifiedName~SqlSchemaCreateNonJsonResponseE2ETests"` (3/3 green on both net8.0 and net10.0).

**Review by Alexandr-Kravchuk — COMMENTED** [review]
## Summary

Lenient re-review round. The verification pass returned no still-open threads of mine on this PR: both earlier Major threads (SchemaNameExists swallowing an unparseable SelectQuery body so SourceCodeSchemaCreate proceeds where it previously aborted, and the missing tests for the rerouted LoadSchema/SaveSchema/EnumerateSchemaLayers parse sites in SchemaDesignerHelper.cs) are already resolved and were not re-raised. The delta pass produced no candidate Blockers, so nothing new reached the objective-Blocker bar. No findings remain; nothing was suppressed. The PR converges — approve.

## Requirement Coverage

re-review round — coverage as per the first review

## Decision

Approve. Posted as a comment because the review runs under the PR author's own account and a self-approval is not permitted.

**Review by m-dymytrova — APPROVED** [review]
## Decision: approve (minor issues)

The core fix is correct and well-evidenced. Verified against the merge-base (`a9deb32`) that the old `SchemaNameExists` was `(uId, _) => uId != null` — it returned "absent" on *any* failure, so a permission error or an empty `SelectQuery` body licensed a create over an existing schema. Replacing it with the discriminated `SchemaResolveStatus` closes a real duplicate-create hole, and `TryCreate_ShouldAbort_WhenDuplicateCheckReturnsFailureEnvelope` specifically pins the "prose happens to say 'not found'" trap. Message classification, the HTML non-echo, and the URL-outside-parens change (I confirmed `SensitiveErrorTextRedactor`'s URI pattern runs to the next whitespace) are all sound. Docs, help, knowledge records and E2E coverage are present.

All findings below are Minor; none blocks merge.

### Validation run (Windows)

```
dotnet test clio.tests/clio.tests.csproj --filter "Category=Unit&(Module=Command|Module=Package)"
→ 4228 passed, 2 failed, 13 skipped
```

The two failures are `OutputPathConfinementTests.Resolve_ShouldFailClosed_OnSymlinkCycle` and
`Resolve_ShouldReject_DanglingIntermediateSymlinkEscape` — symlink-privilege dependent, in files this
PR does not touch, unrelated to the change. Every new test in the diff passed. Per the repo's
smart-regression policy the two changed modules (`Command`, `Package`, no `Common/`) do not trigger
the full suite, so this filter is the right gate. The MCP E2E fixture was read, not executed.

---

## Findings

### Minor — `install-sql-schema` still parses `ScriptSchemaDesignerService` with a bare `JObject.Parse`

`clio/Command/SqlSchemaInstall.cs:70`

`ExecuteScript` on `ServiceModel/ScriptSchemaDesignerService.svc` — the same `.svc` this PR proves can
answer non-JSON — is parsed unguarded, so it can still surface the exact
`Error reading JObject from JsonReader` text this issue was filed about.

The reachability precondition narrows this a lot, and is worth stating: line 70 is only reached when
`ResolveSchemaUId` returns `Resolved`, because both `NotFound` and `Unanswerable` carry a non-null
`Error` and bail at line 61. On the measured stand — no `ScriptSchemaManager` at all — it is
unreachable. It *is* reachable on a build that serves the route and holds a SQL-script schema, when
the call answers with a login page (expired session) or an HTML 500.

Fix: route it through the `ParseServiceResponse` helper this PR introduces, with
`DesignerOperation(Kind, "ExecuteScript")` and `DesignerServiceHint`.

### Minor — read-back verification is applied to `create-sql-schema` only

`clio/Command/SourceCodeSchemaCreate.cs:99`, `clio/Command/SqlSchemaUpdate.cs:96`,
`clio/Command/SourceCodeSchemaUpdate.cs:74`, `clio/Command/ClientUnitSchemaOptions.cs:168`

`create-sql-schema` resolves an `outcomeUnknown` save into an observed outcome via
`VerifyUnknownSaveOutcome`; the other four callers only append `SaveOutcomeUnknownNote` and report a
failure they did not observe. `ResolveSchemaUId` / `LoadSchema` are equally available to all of them,
and the three update commands already hold the schema UId, so the read-back is cheaper there than in
create. Either lift the verification into a shared helper, or say in the PR body why
`create-sql-schema` is the only caller that earns it.

### Minor — the remaining `SchemaNameExists` duplicate-create holes are not in the follow-up list

`clio/Command/ClientUnitSchemaCreate.cs:158`, `clio/Command/PageSchemaMetadataHelper.cs:208`,
`clio/Command/ClientUnitSchemaCreate.cs:196`

Both remaining `SchemaNameExists` variants discard the error
(`(JToken row, _) = QuerySysSchemaRow(...)`) and return `row != null` — the exact fail-open this PR
closes for SQL/SourceCode, still live in `create-client-unit-schema` and `create-page`.
`ClientUnitSchemaCreate.TrySaveSchema` also has an unguarded `JObject.Parse`.

Out of scope to fix here, but the Follow-ups section names only `SourceCodeSchemaCreate` — which this
PR actually fixed — so the list is now stale. Please point it at these three instead.

### Minor — `BuildEmptyBodyMessage` asserts "the response body was empty" on a path where it wasn't

`clio/Package/ServiceResponseJsonGuard.cs:100`

`Deserialize<T>` routes the deserializes-to-`null` case (body `"null"`) through the same message. The
old wording was vague enough to survive that; the new wording is a specific factual claim that is
false there. Pre-existing drift made sharper — worth a distinct message or a softened clause.

---

## Requirement coverage

Source: the issue's Expected / Suggested correction, plus the PR body.

| Criterion | Status |
|---|---|
| Detect a non-JSON / empty service response instead of throwing the raw deserializer message | **covered** — `TryParseJObject` behind all 6 `SchemaDesignerHelper` parse sites; unit + E2E assert `NotContain("Error reading JObject")` |
| Surface a response excerpt | **covered** for the generic unparseable case (bounded, redacted preview); deliberately withheld for HTML with a documented token-leak rationale; N/A for an empty body |
| Surface the HTTP status code | **missing** — structurally impossible through `IApplicationClient`, which returns only the body. Disclosed in the PR body and the knowledge record, deferred to the follow-up issue. Correctly handled: the message states the status is unknown rather than inventing one |
| Expected: "a SQL script schema is created in the target package" | **not achievable on the reported platform** — no SQL-script manager exists in that build. Explicitly disclosed; this PR makes the failure diagnosable, it does not make the command work there |
| Repo policy: docs / MCP / ClioRing review statements | **covered** — help + docs updated, MCP tool contract verified unchanged and E2E added, ClioRing non-consumption verified by grep |

**Confidence limit:** the success path of `create-sql-schema` is unverified on any live stand (no
environment serves the route). The proxy evidence via `SourceCodeSchemaDesignerService` covers the
shared code path but not the SQL-script route itself.
